### DISCUSSION TITLE
Use a data bridge for events

### DISCUSSION THREAD

**Discussion by bkielbasa** [golang-app/ecommerce#21] [Ideas]
I'll use event-driven architecture in this project. For this purpose, I have to decide which technology to choose for it. I'm considering a few options:

* RabbitMQ
* Kafka
* [nats](https://github.com/nats-io/nats-server)
* [EventStoreDB](https://www.eventstore.com/)

If I'd choose today, I think I'd pick Kafka because I know it more but I didn't make a decision yet.

Nats and EventStoreDB look interesting too but I have 0 experience working with them. I'll build a proper abstraction over the technology so I'll be able to change it in the future easily.

My requirements are:
* I have to be able to run it locally with ease
* should be open-source

Nice to have:
* have support from big cloud providers.

If you have any questions/thoughts/suggestions feel free to comment!

**Comment by conexto** [comment]
You’re already thinking in the right direction 👍
Event-driven first + a clean abstraction layer is exactly how I’d approach this.

Some thoughts from real-world usage:

### Kafka
Kafka is the safest long-term bet.

* Strong durability, ordering, and replay
* Excellent ecosystem (Kafka Streams, Flink, Spark → great for analytics & AI pipelines)
* First-class support from AWS (MSK), GCP, Azure, Confluent
* Open-source and very future-proof

Downside is operational complexity and heavier local setup, but if you expect growth, multiple consumers, or replayable event history, Kafka pays off.

### NATS (+ JetStream)
NATS is surprisingly good, especially for developer experience.

* Very easy to run locally (single binary / Docker)
* Low latency, simple mental model
* JetStream adds persistence and replay
* CNCF project with growing cloud support

If the system is more microservice-communication and real-time focused (not analytics-heavy), NATS is a strong and elegant choice.

### RabbitMQ
Solid and mature, but I see it more as a **message broker** than an event-streaming platform.

* Great for task queues, workflows, retries
* Not ideal if events are long-lived, replayed, or used for analytics/ML

### EventStoreDB
Powerful, but only if you fully commit to event sourcing.

* Excellent if ES/CQRS is a core architectural principle
* Steeper learning curve and stronger architectural constraints
* Overkill if you just want “events” rather than full event sourcing

### About abstraction
One thing to be careful about: Kafka-style logs and NATS/Rabbit semantics are fundamentally different.
You can abstract publishing/consuming and serialization, but replay, ordering, offsets, and consumer groups will always leak through.

A thin ports-and-adapters approach works better than trying to hide everything behind a generic interface.

### My take
* Data/AI/analytics heavy system → **Kafka**
* Lightweight, real-time microservices → **NATS**
* Job queues/workflows → **RabbitMQ**
* Pure event-sourced domain → **EventStoreDB**

Since you already know Kafka and it satisfies your requirements, choosing Kafka is completely reasonable, just be aware you’re accepting complexity upfront.

Happy to discuss trade-offs or architecture details further.
